MoreLinq/SelectAsync.cs

```diff
@@ -0,0 +1,439 @@
+                                        tasks.Add(Select(item.Current, cancellationToken));
+                                }
+                                queue.Add(null);
+                            }
+                            catch (Exception e)
+                            {
+                                cancellationTokenSource.Cancel();
+                                queue.Add(ExceptionDispatchInfo.Capture(e));
+                            }
+                            queue.CompleteAdding();
+                        },
+                        CancellationToken.None,
+                        TaskCreationOptions.DenyChildAttach,
+                        scheduler);
+
+                    var nextKey = 0;
+                    var holds = ordered ? new List<KeyValuePair<int, TResult>>() : null;
+
+                    foreach (var e in queue.GetConsumingEnumerable())
+                    {
+                        (e as ExceptionDispatchInfo)?.Throw();
+                        if (e == null)
+                            break;
+
+                        var r = ((Task<KeyValuePair<int, TResult>>) e).Result;
+
+                        if (holds == null || r.Key == nextKey)
+                        {
+                            // If order does not need to be preserved or the key
+                            // is the next that should be yielded then yield
+                            // the result.
+
+                            yield return r.Value;
+
+                            if (holds != null) // preserve order?
+                            {
+                                // Release withheld results consecutive in key
+                                // order to the one just yielded...
+
+                                var releaseCount = 0;
+
+                                for (nextKey++;
+                                     holds.Count > 0 && holds[0] is KeyValuePair<int, TResult> n
+                                                     && n.Key == nextKey;
+                                     nextKey++)
+                                {
+                                    releaseCount++;
+                                    yield return n.Value;
+                                }
+
+                                holds.RemoveRange(0, releaseCount);
+                            }
+                        }
+                        else
+                        {
+                            // Received a result out of order when order must be
+                            // preserved, so withhold the result by finding out
+                            // where it belongs in the order of results withheld
+                            // so far and insert it in the list.
+
+                            var i = holds.BinarySearch(r, KeyValueComparer<int, TResult>.Default);
+                            Debug.Assert(i < 0);
+                            holds.Insert(~i, r);
```

Comment:
> Lists are pretty bad at insertion in the middle.

Yeah, I'm well aware but this is a first cut to focus on getting the API shape and runtime behaviour right.

> This could be postponed until List has been shown to be a problem

Absolutely! There should be room left for improvement. 😇

Reply:
Why not use [SortedList&lt;T>](https://msdn.microsoft.com/en-us/library/ms132319.aspx) ? It seems to fit the purposes quite well, and it should allow removing some book-keeping code.